libs/libcommon/src/libcommon/cloudfront.py

```diff
@@ -0,0 +1,74 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2024 The HuggingFace Authors.
+
+import datetime
+
+from botocore.signers import CloudFrontSigner
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import padding
+
+from libcommon.utils import get_expires
+
+
+class CloudFront:
+    """
+    Signs CloudFront URLs using a private key.
+
+    References:
+    - https://github.com/boto/boto3/blob/develop/boto3/examples/cloudfront.rst
+    - https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-creating-signed-url-canned-policy.html
+    """
+
+    _expiration_seconds: int
+    _signer: CloudFrontSigner
+
+    def __init__(self, key_pair_id: str, private_key: str, expiration_seconds: int) -> None:
+        """
+        Args:
+            key_pair_id (:obj:`str`): The cloudfront key pair id, eg. "K2JCJMDEHXQW5F"
+            private_key (:obj:`str`): The cloudfront private key, in PEM format
+            expiration_seconds (:obj:`int`): The number of seconds the signed url will be valid for
+        """
+        pk = serialization.load_pem_private_key(private_key.encode("utf8"), password=None, backend=default_backend())
+        p = padding.PKCS1v15()
+        h = hashes.SHA1()  # nosec
+        # ^ bandit raises a warning:
+        #     https://bandit.readthedocs.io/en/1.7.5/blacklists/blacklist_calls.html#b303-md5
+        #   but CloudFront mandates SHA1
+
+        def rsa_signer(message: bytes) -> bytes:
+            return pk.sign(message, p, h)  # type: ignore
+            # ^ ignoring mypy type errors (the setup should work, as long as the private key is RSA)
+
+        self._expiration_seconds = expiration_seconds
+        self._signer = CloudFrontSigner(key_pair_id, rsa_signer)
```

Comment:
We'll have to check (in prod ?) that `CloudFrontSigner` works in multithreading (we use multithreading to upload the images/audio to S3 in parallel in /rows)

If `CloudFrontSigner` is not thread safe it might upload partial data or cause deadlocks.

IMO it should be fine though, but let's keep this in mind

Reply:
> We'll have to check (in prod ?)

in staging :)